### PR TITLE
🚩 Milestone 2: Processing data in models | feature/processingData

### PR DESCRIPTION
**Pull request by Ruthmy**
### Hi there! Ruthmy here. 👋

This is the PR for the [Processing data in models](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/Processing_data_model_project_v1_1.md) milestone project. The goal of this milestone is to set up the models and associations for the blog app.

### Here is a summary of what has been done:

- [x] Updated the migration files to match Rails name conventions.
- [x] Updated the `schema.rb` file to match the new migration files.
- [x] Created the following models:
  - [x] User
  - [x] Post
  - [x] Like
  - [x] Comment
- [x] Created the associations between the models.
- [x] Created the following methods:
  - [x] `post_recent` in User, that returns the 3 most recent posts
  - [x] `update_user_posts_counter` in Post,  that updates the posts counter for a user.
  - [x] `recent_comments` in Post, that returns the 5 most recent comments for a post.
  - [x] `update_post_comments_counter` in Comment, that updates the comments counter for a post.
  - [x] `update_post_likes_counter` in Like, that updates the likes counter for a post.

---

### Thank you for taking the time to review this PR ⭐!

---

### CODE REVIEW

File: app/models/post.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+#  likes_counter    :integer
+#
+class Post < ApplicationRecord
+  belongs_to :author, class_name: 'User', foreign_key: :author_id
+  has_many :comments, foreign_key: :post_id
+  has_many :likes, foreign_key: :post_id
+
+  def update_user_posts_counter
+    author.update(posts_counter: author.posts.count)
+  end
```
Comment (DeliceLydia) [inline app/models/post.rb]: ## Optional
- [x] I suggest that you use one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - after_save to call the `#update_user_posts_counter` method every time a new post has been saved to the database, then make the `update_user_posts_counter` method private

File: app/models/like.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+class Like < ApplicationRecord
+  belongs_to :user, foreign_key: :user_id
+  belongs_to :post, foreign_key: :post_id
+
+  after_create :update_post_likes_counter
+  after_destroy :update_post_likes_counter
+
+  def update_post_likes_counter
+    post.update(likes_counter: post.likes.count)
+  end
```
Comment (DeliceLydia) [inline app/models/like.rb:22]: ## Optional
- [x] I suggest that you use one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - after_save to call the `#update_posts_likes_counter` method every time a new `like` has been saved to the database, then make the `update_posts_likes_counter` method private

File: app/models/comment.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  belongs_to :post, foreign_key: :post_id
+
+  validates :text, presence: true
+
+  after_create :update_post_comments_counter
+  after_destroy :update_post_comments_counter
+
+  def update_post_comments_counter
+    post.update(comments_counter: post.comments.count)
+  end
```
Comment (DeliceLydia) [inline app/models/comment.rb:25]: ## Optional
- [x] I suggest that you use one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - after_save to call the `#update_posts_comments_counter` method every time a new `comment` has been saved to the database, then make the `update_posts_comments_counter` method private

**Review by DeliceLydia — APPROVED** [review]
## Hi @Ruthmy, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
# Congratulations! 🎉 

## Highlights
- The post model has a method that returns the 5 most recent comments for a given post
- The Post model has a method that updates the posts counter for a user
- The User model has a method that returns 3 most recent posts
- The Comment model has a method that updates the comments counter for a post
- Like model has a method that updates the likes counter for a post



### Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/post.rb (hunk 1 above)
Comment (Ruthmy) [reply app/models/post.rb]: Hi @DeliceLydia 
I really like this idea you share, thank you very much, I will implement this.

File: app/models/like.rb (hunk 2 above)
Comment (Ruthmy) [reply app/models/like.rb:22]: Implemented the change. Thanks again @DeliceLydia

File: app/models/comment.rb (hunk 3 above)
Comment (Ruthmy) [reply app/models/comment.rb:25]: Implemented the change. Thanks again @DeliceLydia

**Pull request merged**